src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -603,25 +603,50 @@ private void RaiseRequestEvent(HttpWebRequest request)
                     activity.Start();
                 }
 
-                request.Headers.Add(RequestIdHeaderName, activity.Id);
-                // we expect baggage to be empty or contain a few items
-                using (IEnumerator<KeyValuePair<string, string>> e = activity.Baggage.GetEnumerator())
+                if (activity.IdFormat == ActivityIdFormat.W3C)
                 {
-                    if (e.MoveNext())
+                    // do not inject header if it was injected already 
+                    // perhaps tracing systems wants to override it
+                    if (request.Headers[TraceParentHeaderName] == null)
                     {
-                        StringBuilder baggage = new StringBuilder();
-                        do
+                        request.Headers.Add(TraceParentHeaderName, activity.Id);
+                        if (activity.TraceStateString != null)
```

Comment:
Accessing TraceStateString looks like an O(N) operation.  Seems like it should be cached into a local rather than accessing it twice, once to check for null and the other to add it.